### PR TITLE
Allow access to ``_slot`` for local storage pointer variables.

### LINKED ISSUE #8361: Allow assigning to storage pointer variables from inline assembly
References https://github.com/ethereum/solidity/issues/8353#issuecomment-589077551

Restricting access to storage pointer variables but not memory pointer variables is asymmetric. Lifting the restriction would allow some interesting patterns.

### PR DESCRIPTION
**Pull request by chriseth**
Closes https://github.com/ethereum/solidity/issues/8361

### CODE REVIEW

**Comment by chriseth** [comment]
The logic of ControlFlowBuilder / ControlFlowAnalyzer should be extended to inline assembly for this to properly work.

**Comment by leonardoalt** [comment]
Is that going to be done in this PR?

File: docs/assembly.rst [hunk 1]
```diff
@@ -172,6 +172,11 @@ Assignments are possible to assembly-local variables and to function-local
 variables. Take care that when you assign to variables that point to
 memory or storage, you will only change the pointer and not the data.
 
+You can assign to the ``_slot`` part of a local storage variable pointer.
+For these (structs, arrays or mappings), the ``_offset`` part is always zero.
+It is not possible to assign to the ``_slot` or `_offset`` part of a state variable,
```
Comment (leonardoalt) [inline docs/assembly.rst]:
```suggestion
It is not possible to assign to the ``_slot`` or ``_offset`` part of a state variable,
```

**Review by leonardoalt — DISMISSED** [review]
Approving because it looks good so far, not sure if you want to merge it already though.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -677,10 +677,18 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
...
+					{
+						m_errorReporter.typeError(_identifier.location, "State variables cannot be assigned to - you have to use \"sstore()\".");
+						return size_t(-1);
+					}
+					else if (ref->second.isOffset)
+					{
+						m_errorReporter.typeError(_identifier.location, "Only _slot can be assigned to.");
+						return size_t(-1);
+					}
 				}
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp:694]:
```suggestion
                    solAssert(ref->second.isSlot, "");
				}
```

**Comment by mudgen** [comment]
To use this change a person might want to first create an uninitialized local storage pointer.   And then set the pointer in the assembly.   It is an error in Solidity to create an uninitialized local storage pointer. Maybe this restriction should be lifted too?

**Comment by chriseth** [comment]
@mudgen indeed, but it is more work. The extended assignment tracking is planned in https://github.com/ethereum/solidity/issues/8375

**Comment by mudgen** [comment]
@chriseth Makes sense. Thanks.

Why can't we assign to the `_slot` or `_offset` part of a state variable?  I think that would be useful.

**Comment by chriseth** [comment]
@mudgen because that is a compile-time static number.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `3fd3ef4153e30a084dbb0ed7532eb217d9726cbf`:
```
Coding style error:
 libsolidity/analysis/TypeChecker.cpp:692: solAssert(ref->second.isSlot, "");

```
Please check that your changes are working as intended.

**Comment by chriseth** [comment]
Updated.

**Review by leonardoalt — APPROVED** [review]

**Comment by mudgen** [comment]
> @mudgen because that is a compile-time static number.

@chriseth Okay, thanks.

**Pull request merged**